### PR TITLE
Upgrade to Babel 6

### PR DESCRIPTION
**Pull request by jridgewell**
Oh boy, this got big. Upgrades to Babel 6, fixes a slew of issues (see the updated tests), and uses recursive visitors to eliminate (almost) all that state you hated. :wink:

### CODE REVIEW

**Comment by thejameskyle** [comment]
oh god

**Comment by thejameskyle** [comment]
oh god

**Comment by jridgewell** [comment]
You'll want to look at https://github.com/babel-plugins/babel-plugin-incremental-dom/pull/33/files?w=1, so you don't go crazy looking at the test changes.

**Comment by jridgewell** [comment]
Ping @thejameskyle. Can take a quick look? I just want to know if I'm on the right track here.

File: package.json [hunk 1]
```diff
@@ -6,23 +6,23 @@
...
+    "babel-plugin-syntax-jsx": "^6.3.13",
+    "babel-preset-es2015": "^6.3.13",
+    "eslint": "^1.0.0",
     "mocha": "^2.2.5"
   },
   "scripts": {
     "build": "babel-plugin build",
     "push": "babel-plugin publish",
-    "test": "mocha --compilers js:babel/register",
+    "test": "mocha --compilers js:babel-core/register",
```
Comment (thejameskyle) [inline package.json]: Probably want to use the `babel-register` module

**Comment by thejameskyle** [comment]
I'm overwhelmed trying to grasp all this code at once, nothing seems blatantly wrong but there's not much more I can say otherwise

**Comment by jridgewell** [comment]
Yah the diff is pretty large. It might be easier to just review the [finished code](https://github.com/jridgewell/babel-plugin-incremental-dom/tree/babel-6/src).

**Pull request merged**